roofit/roofitcore/src/RooAbsData.cxx

```diff
@@ -2330,8 +2331,43 @@ Bool_t RooAbsData::hasFilledCache() const
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+/// give access to internal data TTree (if it exists)
+/// RooDataSet keeps ownership, RooFit is responsible for deleting.
 
 const TTree* RooAbsData::tree() const
 {
-  return _dstore->tree() ;
+  if (dynamic_cast<const RooTreeDataStore*>(_dstore)) {
+    return _dstore->tree();
+  } else {
+    coutW(InputArguments) << "RooAbsData::tree(" << GetName() << ") WARNING: is not of StorageType::Tree. "
+             << "Use export_tree() instead or convert to tree storage." << endl ;
+    return (TTree*)nullptr;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// clone internal data TTree or create it from vector-based storage
+/// transfers ownership, user is responsible for deleting.
+
+TTree* RooAbsData::export_tree() const
+{
+  if (dynamic_cast<const RooTreeDataStore*>(_dstore)) {
```

Comment:
We should not using dynamic_cast<const RooTreeDataStore*> and cast away the constness on the next line. What about `dynamic_cast<RooTreeDataStore*>` and removing the const_cast?